### ISSUE TITLE
Faster and more accurate solve_poly_system

### ISSUE THREAD

**Issue by filip-misun** [sympy/sympy#14799]
So I was reading the source code of the solve_poly_system function. This is a function that finds the
solutions of a system of multivariate polynomial equations. As far as I understand it, the solutions are found by elimination of the variables, i.e. first, we compute the groebner basis with respect to the lexicographic ordering of the monomials and subsequently, backsubstitution is used to collect the solutions.
This method has at least two issues, however. First of all, groebner bases with respect to the lexicographic ordering are often very large and expensive to compute. Second issue is the backsubstitution. On each level of the backsubsitution, an univariate polynomial has to be solved and these solutions are then substituted into the levels above. Most of the time, univarariate polynomials can only be solved numerically, so we always have just an approximate solution. So on each level of the substitution, we accumulate an error and this error grows in each step of backsubstitution. After several steps, this error can be significant.
The first issue can be solved with easy modification. We can compute the groebner basis w.r.t. to the grevlex ordering, which often requires far less computation and then we can use the FGLM algorithm to convert it to groebner basis w.r.t. the lexicographic ordering.
There is however a better method, which solves both problems of the backsubstitution method simultaneously. The problem of finding a solution to a system of polynomial equations can be reduced to finding eigenvectors of a suitable matrix. This reduction still requires computation of a groebner basis, however, we can use whichever ordering of monomials we like. When we find the eigenvectors of this matrix, we obtain all coordinates of the solutions at once. No error is accumulated. This method is described e.g. in "Cox, David A., John Little, and Donal O'shea. Using algebraic geometry.
So I suggest that either the solve_poly_system is rewritten to use the eigenvalue method, or we allow the user to choose the strategy. If you guys think this is a good idea, I would love to implement it. Thank you!

**Comment by skirpichev** [comment]
> Most of the time, univarariate polynomials can only be solved numerically, so we always have just an approximate solution.

That's not true.  For polynomials with integer/rational coefficients - there are implemented methods (see rootisolation module) to solve such problems exactly (through, AFAIK SymPy still uses roots() function instead of using this fact, so some roots may be lost).

Unfortunately, during backsubstitution you may get new univariate equation with algebraic coefficients (see #9366).  Solving such equations currently supported in SymPy only by symbolic methods.  But it shouldn't be too hard to extend existing root isolation code to support algebraic coefficients.  Do I correctly understand you, that your method will not require such improvement?

> we can use the FGLM algorithm to convert it to groebner basis w.r.t. the lexicographic ordering.

This will work only for zero-dimensional systems.  But SymPy's solve_poly_system() currently doesn't support anything else, so you may not consider this as an issue.

**Comment by filip-misun** [comment]
> I think it's worth keeping both strategies, unless one is always better than the other.

In any case, the backsubstitution method isn't implemented very well in solve_poly_system(), in my opinion. If you look at the source code, you can see that the groebner basis is computed from scratch at each step of the backsubstitution. Computing groebner basis can be very expensive and you don't want to do it more than once if it's not necessary. It is quite easy to implement the backsubstitution method in a way that computes the groebner basis only once. I might write a new issue on this topic.

>  But it shouldn't be too hard to extend existing root isolation code to support algebraic coefficients. Do I correctly understand you, that your method will not require such improvement?

Yes, you are right. But I can imagine, that if there are a lot of steps in the backsubstitution, the coefficients in the polynomials can become pretty complex after a few steps, won't they? If speed matters, the user should have the option to allow approximate solutions, in my opinion. The eigenvalue method on the other hand extracts all the roots by computing eigenvectors of a single matrix. Entries of this matrix can be computed exactly, so this method doesn't need to have a support for algebraic coefficients (unless the polynomials on the input already have algebraic coefficients). The solution produced by the eigenvalue method depends on the solver of the eigenvalue problem - if the eigenvalue solver gives exact solution, so will the eigenvalue method.

**Comment by skirpichev** [comment]
> It is quite easy to implement the backsubstitution method in a way that computes the groebner basis only once.

That might be a good improvement on it's own.  I would like to port such algorithm to the Diofant even if it will be restricted only on zero-dimensional case (Diofant can handle positive-dimensional systems too).

> Entries of this matrix can be computed exactly

Is that means they are rationals?
